src/ripple/app/paths/impl/DirectStep.cpp

```diff
@@ -565,21 +571,43 @@ TER DirectStepI::check (StrandContext const& ctx) const
 
     if (ctx.prevStep)
     {
-        if (auto pds = dynamic_cast<DirectStepI const*> (ctx.prevStep))
+        if (auto prevSrc = ctx.prevStep->directStepSrcAcct ())
         {
-            auto const ter = checkNoRipple (
-                ctx.view, pds->src (), src_, dst_, currency_, j_);
+            auto const ter =
+                checkNoRipple (ctx.view, *prevSrc, src_, dst_, currency_, j_);
             if (ter != tesSUCCESS)
                 return ter;
         }
     }
 
-    if (!ctx.seenDirectIssues[0].insert (Issue{currency_, src_}).second ||
-        !ctx.seenDirectIssues[1].insert (Issue{currency_, dst_}).second)
     {
-        JLOG (j_.debug()) << "DirectStepI: loop detected: Index: "
-                        << ctx.strandSize << ' ' << *this;
-        return temBAD_PATH_LOOP;
+        Issue const srcIssue{currency_, src_};
+        Issue const dstIssue{currency_, dst_};
+
+        if (ctx.seenBookOuts.count (srcIssue))
```

Comment:
I think this would be another good place for a paragraph about why these constitute loops..